api/src/main/java/org/apache/iceberg/util/UnicodeUtil.java

```diff
@@ -40,6 +40,17 @@ public static boolean isCharHighSurrogate(char ch) {
    */
   public static CharSequence truncateString(CharSequence input, int length) {
     Preconditions.checkArgument(length > 0, "Truncate length should be positive");
+    return truncateStringUnsafe(length, input);
+  }
+
+  /**
+   * Truncate the input charSequence such that the returned charSequence is a valid unicode string
+   * and the number of unicode characters in the truncated charSequence is lesser than or equal to
+   * length.
+   *
+   * <p>Note that this function does not validate the input
```

Comment:
Because this uses `UTF8String.substring`, this new method is not called. Can you please revert the changes here and in `Truncate`?

Reply:
Reverted. I might open another PR with these changes as right now the `truncateString` call in the truncate transform validates the input length for every input which is unnecessary.

But it's out of scope for this PR. Updated.

Reply:
Opened https://github.com/apache/iceberg/pull/5496 to avoid validating the input on every record in the CharSequence Truncate transform.

But it's out of scope for this PR so removed from here 👍